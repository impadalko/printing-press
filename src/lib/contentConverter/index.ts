import * as HTML from './HTML'

class ParsedLine {
  content = ''
  quoteDepth = 0

  constructor(line: string, quoteDepth = Infinity) {
    if (line === undefined) return
    this.content = line
    this.quoteDepth = 0
    while (this.content.startsWith('>') && this.quoteDepth < quoteDepth) {
      this.content = this.content.replace(/>\s?/, '')
      this.quoteDepth++
    }
  }
}

const getNextHTMLItem = (
  lines: string[],
  index: number,
  quoteDepth = 0
): [HTML.OptionalHTMLItem, number] => {
  if (index >= lines.length) return [null, index]
  let line = new ParsedLine(lines[index])
  if (line.quoteDepth < quoteDepth) return [null, index]
  if (line.quoteDepth > quoteDepth) {
    const quote = new HTML.Quote()
    while (line.quoteDepth > quoteDepth) {
      let subItem: HTML.OptionalHTMLItem
      ;[subItem, index] = getNextHTMLItem(lines, index, quoteDepth + 1)
      quote.extend(subItem)
      line = new ParsedLine(lines[index])
    }
    return [quote, index]
  }
  if (line.content === '') return [null, index + 1]
  if (line.content.startsWith('# ')) return [new HTML.H1(line.content), index + 1]
  if (line.content.startsWith('## ')) return [new HTML.H2(line.content), index + 1]
  if (line.content.startsWith('### ')) return [new HTML.H3(line.content), index + 1]
  if (line.content.startsWith('#### ')) return [new HTML.H4(line.content), index + 1]
  if (line.content.startsWith('##### ')) return [new HTML.H5(line.content), index + 1]
  if (line.content.startsWith('###### ')) return [new HTML.H6(line.content), index + 1]
  if (line.content === '---') return [new HTML.HR(), index + 1]
  if (line.content === '```') {
    const code = new HTML.Code()
    line = new ParsedLine(lines[++index], quoteDepth)
    while (index < lines.length && line.content != '```') {
      code.extend(line.content)
      line = new ParsedLine(lines[++index], quoteDepth)
    }
    return [code, index + 1]
  }
  const p = new HTML.P(line.content)
  index++
  while (index < lines.length) {
    // This can become inneficient on badly formatted texts
    const [subItem, nextIndex] = getNextHTMLItem(lines, index, quoteDepth)
    if (subItem instanceof HTML.P) {
      p.extend(subItem)
      index = nextIndex
    } else {
      return [p, index]
    }
  }
  return [p, index]
}

export const convert = (lines: string[]): string => {
  let converted = ''
  let nextHTMLItem: HTML.OptionalHTMLItem = null
  let index = 0
  while (index < lines.length) {
    // eslint-disable-next-line @typescript-eslint/no-extra-semi
    ;[nextHTMLItem, index] = getNextHTMLItem(lines, index)
    if (nextHTMLItem !== null) converted += nextHTMLItem.getParsed() + '\n'
  }
  return converted
}
