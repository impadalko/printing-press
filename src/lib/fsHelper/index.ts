import { promises as fs } from 'fs'
import path from 'path'

export const walk = async (inputPath: string): Promise<string[]> => {
  const dirContent = await fs.readdir(inputPath, { withFileTypes: true })
  let files: string[] = []
  for (const entry of dirContent) {
    const entryPath = path.join(inputPath, entry.name)
    if (entry.isDirectory()) files = files.concat(await walk(entryPath))
    else files.push(entryPath)
  }
  return files
}

export const ensureDir = async (filePath: string): Promise<void> => {
  await fs.mkdir(path.dirname(filePath), { recursive: true })
}

export const exists = async (filePath: string): Promise<boolean> => {
  try {
    await fs.access(filePath)
    return true
  } catch {
    return false
  }
}
