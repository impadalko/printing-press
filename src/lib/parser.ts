import fs from 'fs'
import readline from 'readline'
import YAML from 'yaml'

export interface ParsedFile {
  header: unknown;
  content: string[];
}

class ParsingError extends Error {
  constructor(msg: string) {
    super(`Could not parse: ${msg}`)
    Error.captureStackTrace(this, ParsingError)
  }
}

export const parseFile = async (path: string, headerDelimiter = '+++'): Promise<ParsedFile> => {
  const fileStream = fs.createReadStream(path)
  const linesInterface = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  })

  const lines = []
  for await (const line of linesInterface)
    lines.push(line)

  let header = ''
  const closedHeader = lines.some((line, index) => {
    if(index === 0 && line !== headerDelimiter)
      throw new ParsingError('First line must be a header delimiter')
    if(index !== 0 && line === headerDelimiter)
      return true
    if(line !== headerDelimiter){
      header += `${line}\n`
      return false
    }
  })

  if(!closedHeader)
    throw new ParsingError('Header was never closed')

  let parsedHeader = null
  try {
    parsedHeader = YAML.parse(header)
  } catch (e) {
    throw new ParsingError('Header is not a valid YAML')
  }

  let delimiterCount = 0
  const content: string[] = []
  lines.forEach((line) => {
    if (delimiterCount < 2){
      if (line === headerDelimiter)
        delimiterCount++
    }
    else
      content.push(line)
  })

  return {
    header: parsedHeader,
    content: content
  }
}
