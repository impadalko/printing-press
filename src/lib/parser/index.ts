import fs from 'fs'
import readline from 'readline'

type Header = { [k: string]: string }

interface ParsedFile {
  header: Header | null
  content: string[]
}

class ParsingError extends Error {
  constructor(msg: string) {
    super(`Could not parse: ${msg}`)
    Error.captureStackTrace(this, ParsingError)
  }
}

const parseHeader = (header: string): Header | null => {
  if (!header) return null
  const lines = header.split('\n')
  const parsedHeader: Header = {}
  lines.forEach((line) => {
    if (!line) return
    const splittedLine = line.split(':')
    if (splittedLine.length !== 2) throw new ParsingError(`Line "${line}" is not valid in header`)
    parsedHeader[splittedLine[0].trim()] = splittedLine[1].trim()
  })
  return parsedHeader
}

export const parseFile = async (path: string, headerDelimiter = '+++'): Promise<ParsedFile> => {
  const fileStream = fs.createReadStream(path)
  const lines = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity,
  })

  let delimiterCount = 0
  let header = ''
  const content: string[] = []
  for await (const line of lines) {
    if (delimiterCount === 0 && line !== headerDelimiter)
      throw new ParsingError('First line must be a header delimiter')
    if (delimiterCount < 2) {
      if (line === headerDelimiter) delimiterCount++
      else header += `${line}\n`
      continue
    }
    content.push(line)
  }

  if (delimiterCount === 1) throw new ParsingError('Header was never closed')

  const parsedHeader = parseHeader(header)
  return {
    header: parsedHeader,
    content: content,
  }
}
