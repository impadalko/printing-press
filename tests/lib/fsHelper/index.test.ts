import { promises as fs } from 'fs'
import path from 'path'

import { exists, walk, ensureDir } from '../../../src/lib/fsHelper'

test('exists returns true when file exists', () => {
  return exists(`${__dirname}/testFiles/index.txt`).then((data) => {
    expect(data).toBe(true)
  })
})

test('exists returns false when file does not exist', () => {
  return exists(`${__dirname}/testFiles/notIndex.txt`).then((data) => {
    expect(data).toBe(false)
  })
})

test('walk returns all files in a directory recursively', () => {
  const basePath = `${__dirname}/testFiles`
  return walk(basePath).then((data) => {
    const relativePaths = data.map((e) => path.relative(basePath, e))
    expect(relativePaths).toMatchSnapshot()
  })
})

test('walk returns empty array on non-existent folder', () => {
  const basePath = `${__dirname}/nonExistentFolder`
  return walk(basePath).then((data) => {
    expect(data).toEqual([])
  })
})

describe('ensureDir', () => {
  const topDir = `${__dirname}/testFiles/ensureDir`

  afterEach(() => {
    return fs.rm(topDir, { recursive: true, force: true })
  })

  test('ensureDir creates a directory recursively', () => {
    const bottomDir = path.join(topDir, 'a/b/c/')
    const file = path.join(bottomDir, 'index.txt')
    return ensureDir(file).then(() => {
      exists(bottomDir).then((data) => {
        expect(data).toBe(true)
      })
    })
  })
})
