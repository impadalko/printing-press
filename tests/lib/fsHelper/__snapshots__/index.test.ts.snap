// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`walk returns all files in a directory recursively 1`] = `
Array [
  "/home/ivan/Dev/impadalko/printing-press/tests/lib/fsHelper/testFiles/a/index.txt",
  "/home/ivan/Dev/impadalko/printing-press/tests/lib/fsHelper/testFiles/a/x/index.txt",
  "/home/ivan/Dev/impadalko/printing-press/tests/lib/fsHelper/testFiles/b/y/index.txt",
  "/home/ivan/Dev/impadalko/printing-press/tests/lib/fsHelper/testFiles/index.txt",
]
`;
